### PR TITLE
[FEATURE] Implemented removal methods for AzureAiSearchEmbeddingStore

### LINKED ISSUE #1467: [FEATURE] AzureAiSearchEmbeddingStore - "remove" operation
Hello!

I wonder when `AzureAiSearchEmbeddingStore` is going to implement the `EmbeddingStore#remove` and `EmbeddingStore#removeAll` operations.

Thank you!

### PR DESCRIPTION
**Pull request by Devansh-Rastogi**
implemented remove(String id), removeAll(), removeAll(Collection ids), removeAll(Filter filter) methods for AzureAiSearchEmbeddingStore

## Issue
Fixes #1467 

## Change
- implemented remove(String id), removeAll(), removeAll(Collection ids), removeAll(Filter filter) methods for AzureAiSearchEmbeddingStore.
- Added awaitUntilPersisted() after remove methods are called.
- Created AzureAiSearchEmbeddingStoreRemovalIT class to test the changes done

## General checklist

- [x] There are no breaking changes
- [x] I have added unit and integration tests for my change
- [x] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)

### CODE REVIEW

File: langchain4j-azure-ai-search/src/test/java/dev/langchain4j/store/embedding/azure/search/AzureAiSearchEmbeddingStoreRemovalIT.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Arrays.asList;
+import static org.assertj.core.api.Assertions.assertThat;
+
+@EnabledIfEnvironmentVariable(named = "AZURE_SEARCH_ENDPOINT", matches = ".+")
+public class AzureAiSearchEmbeddingStoreRemovalIT extends EmbeddingStoreWithRemovalIT {
```
Comment (langchain4j) [inline langchain4j-azure-ai-search/src/test/java/dev/langchain4j/store/embedding/azure/search/AzureAiSearchEmbeddingStoreRemovalIT.java:14]: Please add the same test for `AzureAiSearchContentRetriever` (`AzureAiSearchContentRetrieverRemovalIT`)

File: langchain4j-azure-ai-search/src/test/java/dev/langchain4j/store/embedding/azure/search/AzureAiSearchEmbeddingStoreRemovalIT.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+    private static final Logger log = LoggerFactory.getLogger(AzureAiSearchEmbeddingStoreIT.class);
```
Comment (langchain4j) [inline langchain4j-azure-ai-search/src/test/java/dev/langchain4j/store/embedding/azure/search/AzureAiSearchEmbeddingStoreRemovalIT.java]: nit: wrong class in the logger

File: langchain4j-azure-ai-search/src/test/java/dev/langchain4j/store/embedding/azure/search/AzureAiSearchEmbeddingStoreRemovalIT.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+    private EmbeddingModel embeddingModel;
+
+    private EmbeddingStore<TextSegment> embeddingStore;
+
+    private int dimensions;
```
Comment (langchain4j) [inline langchain4j-azure-ai-search/src/test/java/dev/langchain4j/store/embedding/azure/search/AzureAiSearchEmbeddingStoreRemovalIT.java]: minor: no need to store `dimensions` in the variable, you can always get it from `embeddingModel.dimension()`

File: langchain4j-azure-ai-search/src/test/java/dev/langchain4j/store/embedding/azure/search/AzureAiSearchEmbeddingStoreRemovalIT.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    }
+
+    @BeforeEach
+    void setUp() {
+        clearStore();
+    }
+
+
+    @Test
+    void test_add_embeddings_and_remove_document_and_find_relevant(){
```
Comment (langchain4j) [inline langchain4j-azure-ai-search/src/test/java/dev/langchain4j/store/embedding/azure/search/AzureAiSearchEmbeddingStoreRemovalIT.java]: Are these (this and 2 below) tests really required? Usually tests inherited from `EmbeddingStoreWithRemovalIT` should be enough

File: langchain4j-azure-ai-search/src/main/java/dev/langchain4j/store/embedding/azure/search/AbstractAzureAiSearchEmbeddingStore.java [hunk 5]
```diff
@@ -263,6 +264,48 @@ public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedde
...
+
+    @Override
+    public void removeAll(){
+        SearchOptions searchOptions = new SearchOptions().setSelect(DEFAULT_FIELD_ID);
+        SearchPagedIterable searchResults = searchClient.search("*", searchOptions, Context.NONE);
+        List<String> ids = searchResults.stream()
+                .map(searchResult -> searchResult.getDocument(SearchDocument.class))
+                .map(doc -> (String) doc.get(DEFAULT_FIELD_ID))
+                .collect(Collectors.toList());
+        removeAll(ids);
```
Comment (langchain4j) [inline langchain4j-azure-ai-search/src/main/java/dev/langchain4j/store/embedding/azure/search/AbstractAzureAiSearchEmbeddingStore.java:294]: Q: are there any limits on the number of documents that can be removed at once?

**Review by ghost — COMMENTED** [review]
@Devansh-Rastogi thank you!

File: langchain4j-azure-ai-search/src/test/java/dev/langchain4j/store/embedding/azure/search/AzureAiSearchEmbeddingStoreRemovalIT.java (hunk 1 above)
Comment (Devansh-Rastogi) [reply langchain4j-azure-ai-search/src/test/java/dev/langchain4j/store/embedding/azure/search/AzureAiSearchEmbeddingStoreRemovalIT.java:14]: added AzureAiSearchContentRetrieverRemovalIT

File: langchain4j-azure-ai-search/src/test/java/dev/langchain4j/store/embedding/azure/search/AzureAiSearchEmbeddingStoreRemovalIT.java (hunk 2 above)
Comment (Devansh-Rastogi) [reply langchain4j-azure-ai-search/src/test/java/dev/langchain4j/store/embedding/azure/search/AzureAiSearchEmbeddingStoreRemovalIT.java]: corrected

File: langchain4j-azure-ai-search/src/test/java/dev/langchain4j/store/embedding/azure/search/AzureAiSearchEmbeddingStoreRemovalIT.java (hunk 3 above)
Comment (Devansh-Rastogi) [reply langchain4j-azure-ai-search/src/test/java/dev/langchain4j/store/embedding/azure/search/AzureAiSearchEmbeddingStoreRemovalIT.java]: removed occurences of dimensions and now using embeddingModel.dimension()

File: langchain4j-azure-ai-search/src/test/java/dev/langchain4j/store/embedding/azure/search/AzureAiSearchEmbeddingStoreRemovalIT.java (hunk 4 above)
Comment (Devansh-Rastogi) [reply langchain4j-azure-ai-search/src/test/java/dev/langchain4j/store/embedding/azure/search/AzureAiSearchEmbeddingStoreRemovalIT.java]: not really required.. removed them

File: langchain4j-azure-ai-search/src/main/java/dev/langchain4j/store/embedding/azure/search/AbstractAzureAiSearchEmbeddingStore.java (hunk 5 above)
Comment (Devansh-Rastogi) [reply langchain4j-azure-ai-search/src/main/java/dev/langchain4j/store/embedding/azure/search/AbstractAzureAiSearchEmbeddingStore.java:294]: in documentation there is nothing on limits for the number of documents, however i see SearchClient class uses
SearchAsyncClient.buildIndexBatch before deleting the docs

**Comment by Devansh-Rastogi** [comment]
implemented the PR review comments

**Review by ghost — APPROVED** [review]
@Devansh-Rastogi thank you, good job!

File: langchain4j-azure-ai-search/src/main/java/dev/langchain4j/store/embedding/azure/search/AbstractAzureAiSearchEmbeddingStore.java [hunk 6]
```diff
@@ -263,6 +264,48 @@ public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedde
...
+    public void removeAll(Collection<String> ids){
+        ensureNotEmpty(ids, "ids");
+        List<Map<String, String>> documentsToDelete = new ArrayList<>();
+        for (String id : ids) {
+            ensureNotBlank(id, "id");
+            Map<String, String> documents = new HashMap<>();
+            documents.put(DEFAULT_FIELD_ID, id);
+            documentsToDelete.add(documents);
+        }
+        searchClient.deleteDocuments(documentsToDelete);
```
Comment (vprudnikov) [inline langchain4j-azure-ai-search/src/main/java/dev/langchain4j/store/embedding/azure/search/AbstractAzureAiSearchEmbeddingStore.java:283]: Should we inspect the result of the operation?
For example:
```
        var indexDocumentsResult = searchClient.deleteDocuments(documentsToDelete);
        var errorMessage = indexDocumentsResult.getResults().stream()
                .filter(not(IndexingResult::isSucceeded))
                .reduce("", (acc, r) -> acc +
                                "Document %s could not be deleted. Status: %s , error: %s\n".formatted(r.getKey(), r.getStatusCode(), r.getErrorMessage()),
                        String::concat);
        if (!errorMessage.isEmpty()) {
            throw new AzureException(errorMessage);
        }
```

**Comment by vprudnikov** [comment]
I added a comment for https://github.com/langchain4j/langchain4j/pull/1563/files#diff-06626224ac1d5b77af8075520d1a9a3a179f5cc20e544fa37833e9bdf28ff9e9

Not sure though how to ideomatically use the SearchClient.

File: langchain4j-azure-ai-search/src/main/java/dev/langchain4j/store/embedding/azure/search/AbstractAzureAiSearchEmbeddingStore.java (hunk 6 above)
Comment (langchain4j) [reply langchain4j-azure-ai-search/src/main/java/dev/langchain4j/store/embedding/azure/search/AbstractAzureAiSearchEmbeddingStore.java:283]: good point!

**Pull request merged**